src/RECONR/ReferenceGrid/src/fill.hpp

```diff
@@ -0,0 +1,14 @@
+static void fill( const double lowerBound,
+                  const double upperBound,
+                  std::vector<double>& buffer ){
+  constexpr double thresholdRatio = 1.193776641714437;
+  if ( lowerBound * thresholdRatio < upperBound ){
```

Comment:
I think I know what this is trying to do (add logarithmically spaced energy values between `lowerBound` and `upperBound`), but I can't quite follow what is happening. The syntax is a bit obscure.

Reply:
let's unpack it one step at a time.

```cpp
[ =, state = lowerBound ]() mutable { return state *= thresholdRatio; }
```

This is a mutable lambda. Each time it's invoked, it generates a different value, i.e.
```
call -> lowerBound
call -> lowerBound * thresholdRatio
call -> lowerBound * thresholdRatio * thresholdRatio
call -> lowerBound * thresholdRatio * thresholdRatio * thresholdRatio
...
```

`ranges::view::generate` takes a callable objects and returns an infinite (lazy) range of values from repeatedly calling it's parameter object, e.g.

```
[ lowerBound,
  lowerBound * thresholdRatio,
  lowerBound * thresholdRatio * thresholdRatio,
  lowerBound * thresholdRatio * thresholdRatio * thresholdRatio,
  ... ]
```

`ranges::view::take_while` takes values from an input range until a predicate fails. In this case, it takes values while those values are less than the upper bound.

`ranges::action::push_back` is appending each value in the range to buffer.